### ISSUE TITLE
"Hide network" does not work as expected on remote instance

### ISSUE THREAD

**Issue by Kjwon15** [tootsuite/mastodon#11390]
"Hide network" is not specified in AP, so this problem happened.

### Expected behaviour

Some user who checked "hide network" setting, that one's followers/following should not be visible on remote instance.

### Actual behaviour

Remote users' still find out that user's followers/followings in their local instance.
e.g. If a@A follows x@X, b@A can see that on A

### Steps to reproduce the problem

Find some mastodon user that enabled "hide network" setting.
That user's public page is not showing their network.
Check that user on other instance, followers/followings are exposed.

### Specifications

Mastodon version: v2.9.2 (stable)

**Comment by trwnh** [comment]
Each instance is only aware of *local* followers of that account. You can't see all the followers, but you also currently can't hide all the followers.

Fortunately, I think federating this information should be simple -- just don't show the tab if that actor's JSON representation does not have `follower` or `following` properties, and make enabling the "Hide network" feature remove those two properties from the JSON. (ActivityPub says that an actor SHOULD have follower/following collections, not MUST.)

**Comment by TemporaryUsername12481** [comment]
If an actor doesn't have those properties you can throw `sharedInbox` out the window

**Comment by trwnh** [comment]
@TemporaryUsername12481 sharedInbox is completely irrelevant to this. the local server is responsible for processing/redelivering activities it receives there. it is already aware of all subscriptions for local users, and does not need to publish any info to do its job.

**Comment by TemporaryUsername12481** [comment]
The server can't know that the `to` field contains a collection of followers for the actor creating the message unless the actor creating the message exposed which collection is its followers.

**Comment by rinpatch** [comment]
> Fortunately, I think federating this information should be simple -- just don't show the tab if that actor's JSON representation does not have follower or following properties, and make enabling the "Hide network" feature remove those two properties from the JSON. (ActivityPub says that an actor SHOULD have follower/following collections, not MUST.)

How would one address the follower collection if the remote instances don't know at least it's URI?

What Pleroma does is fetch the first page of the follower collection, if it returns a `CollectionPage` or an `OrderedCollectionPage`, it marks it as public, if the server responds with 401/403, it is marked as private.

**Comment by trwnh** [comment]
@TemporaryUsername12481 well hmm it sounds like sharedInbox is fundamentally flawed then :-)))

@rinpatch I think it's not desirable to address follower collections directly, but let's use your described heuristic as an example. If Pleroma received the following in its `sharedInbox`:

```
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "type": "Create",
  "id": "https://trwnh.com/activities/1",
  "actor": "https://trwnh.com/actors/1",
  "object": {
    "id": "https://trwnh.com/objects/1",
    "type": "Note",
    "attributedTo": "https://trwnh.com/actors/1",
    "content": "This is a note",
    "published": "2015-02-10T15:04:55Z",
    "to": ["https://trwnh.com/collections/1"]
  },
  "published": "2015-02-10T15:04:55Z",
  "to": ["https://trwnh.com/collections/1"]
}
```

then it would try to fetch `https://trwnh.com/collections/1`, no? So what would Pleroma do if it response is 401 Not Authorized? 

Let's vary it a little bit and say that `https://trwnh.com/collections/1` resolved successfully, responded with 200 OK, and returned a `Collection` of `Actor` objects. What would Pleroma do then?

Let's say that fetching `https://trwnh.com/actors/1` yields the following:

```
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "type": "Person",
  "id": "https://trwnh.com/actors/1",
  "inbox": "https://trwnh.com/inboxes/1",
  "outbox": "https://trwnh.com/outboxes/1",
  "preferredUsername": "trwnh",
  "name": "a",
  "summary": "hi, it's me, trwnh.",
  "icon": [
    "https://trwnh.com/actors/1/icon.png"
  ]
}
```

So far everything is spec-compliant.

**Comment by rinpatch** [comment]
>  I think it's not desirable to address follower collections directly

From [AP Spec](https://www.w3.org/TR/activitypub/#followers):
>Note: Default for notification targeting
>
>The follow activity generally is a request to see the objects an actor creates. This makes the Followers collection an appropriate default target for delivery of notifications. 

>  Targets for delivery are determined by checking the ActivityStreams audience targeting; namely, the to, bto, cc, bcc, and audience fields of the activity.

Prismo is the only implementation I am aware of, that doesn't address the follower collection in audience fields.

> If Pleroma received the following in its sharedInbox [...] then it would try to fetch https://trwnh.com/collections/1, no?

No, it doesn't try to fetch collections addressed in activities. When (re)fetching the actor, it would look for `following`/`followers`, fetch them, then try fetching the first pages and mark actor's following/followers as private if it receives 401/403 instead of an (Ordered)Collection page.

Also as @TemporaryUsername12481 pointed out, the remote server can't determine if the collection really belongs to the actor, unless the actor exposes it's uri. Can't speak for other implementations, but Pleroma strips all follower collections that don't belong to the actor when processing remote activities for spam protection, so statuses of the actor with no following/followers collection defined won't show up on timelines.

**Comment by trwnh** [comment]
>Targets for delivery are determined by checking the ActivityStreams audience targeting; namely, the to, bto, cc, bcc, and audience fields of the activity.

Yes, but also from the AP spec [(7.1 Delivery)](https://www.w3.org/TR/activitypub/#delivery):

>If a recipient is a Collection or OrderedCollection, then the server MUST dereference the collection (with the user's credentials) and discover inboxes for each item in the collection.

This is actually a MUST, which is stronger than the SHOULD that is used for publishing `followers`.

---

In general, exposing the URI is not, and should not, be a requirement. Otherwise, you could easily address activities to *someone else's followers* as described in [7.1.2 Inbox Forwarding](https://www.w3.org/TR/activitypub/#inbox-forwarding), which a compliant server MUST redeliver, unless it filters the activity as spam (with implementation-specific rules). Suppose I sent all my activities to everyone's `followers` collections. This is a huge vector for spam, which Pleroma apparently bypasses only by virtue of stripping out the audience targeting entirely as you described. It also precludes a more context-aware approach, such as sending a non-public reply addressed to your conversational partner's followers, rather than your own.

However, I do agree that it is probably better overall to use an authorization check, regardless of whether the `followers` property is published or not. And as to addressing arbitrary collections, there is still the issue of whether the dereferencing should be done by the origin server before delivery (as per the spec), or whether it should be possible for the receiving server to dereference as well (in the case of sharedInbox delivery, which is the source of a lot of these issues).

**Comment by ThibG** [comment]
It has been implemented in #11673, but will only trigger when accounts get re-processed after both instances were upgraded.
